halo2_proofs/src/dev.rs

```diff
@@ -915,8 +915,18 @@ impl<F: FieldExt> MockProver<F> {
                     };
 
                     assert!(lookup.table_expressions.len() == lookup.input_expressions.len());
-                    let fill_row = vec![Value::Real(F::zero()); lookup.table_expressions.len()];
-                    let mut table_has_fill_row = false;
+                    assert!(self.usable_rows.end > 0);
+
+                    // We optimize on the basis that the last usable row might have been repeated to fill
+                    // the table (it doesn't matter if it wasn't). Note that this "fill row" necessarily
```

Comment:
This doesn't match what we actually do in either `SimpleFloorPlanner` or `floor_planner::V1`, which is to use the _first_ row of the table assignment as the fill row:

https://github.com/zcash/halo2/blob/221766986b52d2a35f60d94266647d9dcf908231/halo2_proofs/src/circuit/floor_planner/single_pass.rs#L437-L444

As you say, it doesn't really matter; the performance improvement we observe is not due to this optimisation.

Reply:
This is a misunderstanding of the comment. Whichever policy was used to pick the fill row, provided that there were any fill rows, the table contents after filling will have the contents of the fill row in the last usable row. (If there were no fill rows, treating the last usable row as the potential fill contents is still harmless.)